Infer method calls on known types and warn about nonexistent methods

radeusgd opened the pull request:
### Pull Request Description

- Implements #9812 
- Allows us to propagate type inference through method calls, at least in a basic way.
- Adds a 'lint warning': when calling a method on an object of a known specific (non-`Any`) type that does not exist on that type, a warning is reported, because such a call would always result in `No_Such_Method` error at runtime.
	- `Any` has special behaviour - if `x : Any` we don't know what `x` might be, so we allow all methods on it and defer to the runtime to see if they will be valid or not.
	- This check is not proving correctness, but instead it's only triggering for 'provably wrong' situations.
- Includes changes from #11955: 
	- removing obsolete `AtomTypeInterface`, 
	- simplifying `TypeRepresentation` to be a plain data structure, 
	- and removing a cycle from IR in `BindingsMap` in favour of relying on `StaticModuleScope`.

### Important Notes

<!--
- Mention important elements of the design.
- Mention any notable changes to APIs.
-->

### Checklist

Please ensure that the following checklist has been satisfied before submitting the PR:

- [x] The documentation has been updated, if necessary.
- [ ] Screenshots/screencasts have been attached, if there are any visual changes. For interactive or animated visual changes, a screencast is preferred.
- [x] All code follows the
      [Scala](https://github.com/enso-org/enso/blob/develop/docs/style-guide/scala.md),
      [Java](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md),
      [TypeScript](https://github.com/enso-org/enso/blob/develop/docs/style-guide/typescript.md),
      and
      [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md)
      style guides. In case you are using a language not listed above, follow the [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md) style guide.
- [x] Unit tests have been written where possible.
- [ ] If meaningful changes were made to logic or tests affecting Enso Cloud integration in the libraries, 
      or the Snowflake database integration, a run of the [Extra Tests](https://github.com/enso-org/enso/actions/workflows/extra-nightly-tests.yml) has been scheduled.
  - If applicable, it is suggested to paste a link to a successful run of the Extra Tests.

radeusgd committed:
rename pass

radeusgd committed:
WIP: find types from checked signatures

radeusgd committed:
rename CompilerTest -> ParserTest

radeusgd committed:
refactor: move some common methods to separate class to allow for reuse

radeusgd committed:
add a test for the type extraction

radeusgd committed:
fix test

radeusgd committed:
WIP tests

radeusgd committed:
tests for methods

radeusgd committed:
add a test verifying return-type check on member methods - to test my hypothesis

radeusgd committed:
add a test that finds the cause of issue: FORCE node makes the ascription no longer in outer layer

radeusgd committed:
fix resolution of return types in type signatures

radeusgd committed:
adding tests for member method calls

radeusgd committed:
find the right place to resolve calls

radeusgd committed:
use interpreter context to be able to use imports in the type inference tests

Ideally, the static analysis tests should be runnable without the Graal runtime context. We'd like to be able to run these static passes without Graal runtime, e.g. inside of code editors. But refactoring import resolution and PackageRepository to work without Graal is out of scope of the current types work. It can be planned for later, when needed. For now the focus is to make type checking provide actual value - before that any editor integration wouldn't be useful anyway.

radeusgd committed:
javafmt

radeusgd committed:
update tests

radeusgd committed:
fix test

radeusgd committed:
WIP

radeusgd committed:
fix compilation (stub)

radeusgd committed:
better message in test

radeusgd committed:
WIP: StaticModuleScope

radeusgd committed:
WIP: StaticModuleScope 2

radeusgd committed:
WIP

radeusgd committed:
checkpoint

radeusgd committed:
wiring the StaticModuleScopeAnalysis pass with rest of the system

radeusgd committed:
formatting

radeusgd committed:
debugging

radeusgd committed:
WIP: workaround for static methods on module?

radeusgd committed:
debugging

radeusgd committed:
debug

radeusgd committed:
WIP: test

radeusgd committed:
resolving static module methods works

radeusgd committed:
get rid of unnecessary parentheses in type string

radeusgd committed:
javafmt

radeusgd committed:
ensure that test modules are registered in tests

radeusgd committed:
no such method tests

radeusgd committed:
member method call through static syntax

radeusgd committed:
remove some prints

radeusgd committed:
test calling static method on member

radeusgd committed:
WIP: static syntax for member methods

radeusgd committed:
fix

radeusgd committed:
type hierarchy, reporting no such method

radeusgd committed:
fix error locations, less logs

radeusgd committed:
javafmt, add pass

radeusgd committed:
heisenbug?!

radeusgd committed:
change order of compiler passes to have needed data available

radeusgd committed:
remove debug print

radeusgd committed:
adjust missing builtin

radeusgd committed:
fix static call syntax of member methods by correctly infrring their types at the type signature level

radeusgd committed:
add test for future

radeusgd committed:
resolving Self type to the actual type

radeusgd committed:
fix test - metadata is now only attached to bindings

radeusgd committed:
nit

radeusgd committed:
add test for fields access

radeusgd committed:
WIP test for inter-module definitions

radeusgd committed:
WIP: debugging calling methods from different module

radeusgd committed:
adjust warning message

radeusgd committed:
adjust warning message, and other stuff

radeusgd committed:
checkpoint

radeusgd committed:
fix a test

radeusgd committed:
add a test - even if Any is not imported we should know its base methods - method resolution works on the Builtin somehow?

radeusgd committed:
WIP

radeusgd committed:
fix indentation (?)

radeusgd committed:
test local extension method

radeusgd committed:
inferring field getters

radeusgd committed:
fixes after rebase

radeusgd committed:
javafmt

radeusgd committed:
update links in doc

radeusgd committed:
very basic registering of imports

radeusgd committed:
WIP

radeusgd committed:
WIP

radeusgd committed:
basic imports work

radeusgd committed:
javafmt

radeusgd committed:
checkpoint

radeusgd committed:
hardcode builtins for Any

radeusgd committed:
add missing builtin method to make resolution checker happy

radeusgd committed:
figuring out resolution on Functions and duplicated imports?

radeusgd committed:
add missing builtins

radeusgd committed:
remove debug print to avoid spam

radeusgd committed:
fix sigs

radeusgd committed:
fix how we find the package repo (before it was missing in some paths, now it appears everywhere where needed)

radeusgd committed:
WIP: resolving Type 'call' on module - phase 1 not report it as error, TODO: actually resolve it to the type

radeusgd committed:
Merge branch 'develop' into wip/radeusgd/9812-infer-method-calls

radeusgd committed:
javafmt

radeusgd committed:
get it to compile

radeusgd committed:
fix ids after merge

radeusgd committed:
javafmt once more

radeusgd committed:
fixing conflicts, removing empty files

radeusgd committed:
moved classes to fix compilation

radeusgd committed:
fix NPE in test

radeusgd committed:
use right getter

radeusgd committed:
mark private

radeusgd committed:
try using a more random hash merge

radeusgd committed:
add some missing registrations

radeusgd committed:
if --no-ir-cache is set, --compile will not save caches - it will just run a 'lint' pass; if ir caches are enabled, --compile does save

radeusgd committed:
lower log level

radeusgd committed:
Merge branch 'develop' into wip/radeusgd/9812-infer-method-calls

radeusgd committed:
trying to look up all exported methods

radeusgd committed:
Merge branch 'develop' into wip/radeusgd/9812-infer-method-calls

radeusgd committed:
make tests compile

radeusgd committed:
add a test for re-exports

radeusgd committed:
fix re export resolution

radeusgd committed:
find in imports: keep logic in one place

radeusgd committed:
more tests but ignored for now

radeusgd committed:
use Builder for StaticModuleScope

radeusgd committed:
doc

radeusgd committed:
shape for common method resolution

radeusgd committed:
stub

radeusgd committed:
add missing builtin to micro-distribution

radeusgd committed:
use common method resolution algorithm in compiler

radeusgd committed:
fix brackets in tests after improving toString

radeusgd committed:
add missing builtin, missing arg

radeusgd committed:
use common algorithm in runtime method resolution

radeusgd committed:
`private` causes Private_Access errors - out of scope of this PR, so reverting adding private; probably to be resolved in https://github.com/enso-org/enso/issues/6282

radeusgd committed:
Merge branch 'develop' into wip/radeusgd/9812-infer-method-calls

radeusgd committed:
fix after merge

radeusgd committed:
move to better location

radeusgd committed:
WIP trying to extract common logic

radeusgd committed:
WIP 2

radeusgd committed:
integrate the common logic alg with IrToTruffle

radeusgd committed:
WIP some cleanup in missing arg, starting work on static

radeusgd committed:
common parts in static module scope analysis

radeusgd committed:
make it work

radeusgd committed:
fix original common code - there was a typo...

radeusgd committed:
doc

radeusgd committed:
doc

radeusgd committed:
more docs

radeusgd committed:
do not log in tests

radeusgd committed:
logger

radeusgd committed:
remove stale todo

radeusgd committed:
common getExportedMethod

radeusgd committed:
but now properly

radeusgd committed:
common logic for conversions too

radeusgd committed:
remove AtomType stub - it was not finished - to be re-added in next iteration

radeusgd commented:
Benchmarks scheduled:
- engine: https://github.com/enso-org/enso/actions/runs/11559293758
- libs: https://github.com/enso-org/enso/actions/runs/11559302083

<blockquote><img src="https://opengraph.githubassets.com/0e8555556eef8f44c29d3c69cfe44c47cf111f84cea49f70902ce8120d87565b/enso-org/enso" width="48" align="right"><div><img src="https://github.githubassets.com/favicons/favicon.svg" height="14"> GitHub</div><div><strong><a href="https://github.com/enso-org/enso">Benchmark Engine · enso-org/enso@5be3b82</a></strong></div><div>Hybrid visual and textual functional programming. Contribute to enso-org/enso development by creating an account on GitHub.</div></blockquote>
<blockquote><img src="https://opengraph.githubassets.com/0e8555556eef8f44c29d3c69cfe44c47cf111f84cea49f70902ce8120d87565b/enso-org/enso" width="48" align="right"><div><img src="https://github.githubassets.com/favicons/favicon.svg" height="14"> GitHub</div><div><strong><a href="https://github.com/enso-org/enso">Benchmark Standard Libraries · enso-org/enso@5be3b82</a></strong></div><div>Hybrid visual and textual functional programming. Contribute to enso-org/enso development by creating an account on GitHub.</div></blockquote>

radeusgd committed:
Merge branch 'develop' into wip/radeusgd/9812-infer-method-calls

# Conflicts:
#	engine/runtime-compiler/src/main/scala/org/enso/compiler/context/InlineContext.scala

radeusgd committed:
remove assert

radeusgd committed:
make it okay to miss a method pointer - just return null

JaroslavTulach commented on distribution/lib/Standard/Base/0.0.0-dev/src/Data/Ordering.enso:
```diff
@@ -129,6 +129,10 @@ type Comparable
 
+    ## PRIVATE
+    less_than_builtin : Any -> Any -> Boolean | Nothing
+    less_than_builtin = @Builtin_Method "Comparable.less_than_builtin"
```
Please remove this method. `@Builtin_Method`s are supposed to be only in `private` modules. Moreover there already is such a `less_than_builtin` in `Ordering_Helpers.enso`.

radeusgd replied:
Ok, this method is indeed wrong, as I don't see a `Comparable.less_than_builtin` builtin in Java.

It was probably an artifact from some merges that happened along the way (or just my oversight). I'll remove this one.

JaroslavTulach commented on distribution/lib/Standard/Base/0.0.0-dev/src/Internal/Ordering_Helpers.enso:
```diff
@@ -37,8 +37,10 @@ type Default_Comparator
     ## PRIVATE
     hash : Number -> Integer
     hash x = Default_Comparator.hash_builtin x
+
```
Unnecessary changes.

radeusgd replied:
I'm just trying to apply consistent formatting.

JaroslavTulach commented on distribution/lib/Standard/Base/0.0.0-dev/src/Meta/Enso_Project.enso:
```diff
@@ -91,6 +91,9 @@ type Project_Description
     namespace self = self.ns
 
+    ## PRIVATE
+    enso_project_builtin module = @Builtin_Method "Project_Description.enso_project_builtin"
```
Please don't put `Builtin_Method`s into non-`private` modules.

Akirathan replied:
@radeusgd has a good point. The corresponding builtin method is already defined as [EnsoProjectNode](https://github.com/enso-org/enso/blob/f0e9616d39d4fc24c8074bd4681fed182d0ede00/engine/runtime/src/main/java/org/enso/interpreter/node/expression/builtin/meta/EnsoProjectNode.java). Thus, this code does not introduce any changes. It seems that [BuiltinFunction.isAutoRegister](https://github.com/enso-org/enso/blob/7b0759f8b3b7ab4276a8668642dbd4b9586fafdd/engine/runtime/src/main/java/org/enso/interpreter/runtime/builtin/BuiltinFunction.java#L16) makes it possible to defined a builtin method in Java without any *shadow* definition in Enso code. I think having these kind of *shadow* definitions is beneficial, as we can immediately see all the definitions of builtin methods on a type.

hubertp replied:
Yes, that's how auto register was supposed to work.

JaroslavTulach commented on distribution/lib/Standard/Base/0.0.0-dev/src/System/File.enso:
```diff
@@ -832,6 +832,12 @@ type File
     to_display_text self = self.to_text
 
+    ## PRIVATE
+    copy_builtin self target options = @Builtin_Method "File.copy_builtin"
```
Seeing all these (undesirable) `@Builtin_Method`s changes, I wonder how are they related to _type inference_?

radeusgd replied:
There are no _changes_ to builtin methods structure.

I'm only reproducing in Enso code the structure that **has already been there** in the builtins. That is - we already were having calls to `File.copy_builtin` or `Comparable.less_than_builtin` in the code, they were just lacking the related definitions.

If that structure is _undesirable_, we should restructure the builtins, but that is not really in the scope of the PR. But if you will not allow me to merge it without that, I will probably be forced to do this _unrelated_ change.


> I wonder how are they related to _type inference_?

I've added checking of method calls. What methods are available on an object was taken from the type definitions. If a builtin method was called e.g. `File.copy_builtin` but there was no `copy_builtin` defined on `File`, my type inference was giving warnings about calling nonexstent methods. I don't think builtins shall be a special case (apart from a naming convention I cannot easily tell if a method is a builtin or doesn't exist). Most of our builtins already had such 'shadow definitions', there was just a few that were lacking and I've added these missing ones. See for example, `File.input_stream_builtin` - it's been there for a long time. I think we should do a proper refactor of builtins to move them into private modules. But that's the #6282 ticket.

radeusgd replied:
On the other hand, these changes are not absolutely **required** for the type inference changes. They are there to make the tests pass, but I can `@Ignore` a few tests and revert these changes so that we can handle them separately. Maybe I should have done them as a separate PR like with #11422

Akirathan replied:
I like these kind of *shadow* definitions of builtin methods. They don't (at least they should not) introduce any changes (that is, they don't introduce any new builtins) and they make it easy to see which builtin methods are defined on a type. It may be worth revisiting [BuiltinMethod.autoRegister](https://github.com/enso-org/enso/blob/c2c6712b77cf8fc7aba8a478e6169ee1dceb23ce/lib/scala/interpreter-dsl/src/main/java/org/enso/interpreter/dsl/BuiltinMethod.java#L41) annotation argument.

See [my other comment](https://github.com/enso-org/enso/pull/11399#discussion_r1827559455)

JaroslavTulach replied:
> If a builtin method was called e.g. `File.copy_builtin` but there was no `copy_builtin` defined on `File`, my type inference was giving warnings about calling nonexstent methods. 

I see. That's indeed related.

> I don't think builtins shall be a special case 

Builtin methods are _special case_! They are not behaving as normal methods. 

> we should do a proper refactor of builtins to move them into private modules. ... that's the #6282 ticket.

> ... shadow definitions of builtin methods ... They don't introduce any changes ...

OK.

> worth revisiting [BuiltinMethod.autoRegister](https://github.com/enso-org/enso/blob/c2c6712b77cf8fc7aba8a478e6169ee1dceb23ce/lib/scala/interpreter-dsl/src/main/java/org/enso/interpreter/dsl/BuiltinMethod.java#L41) annotation argument.

- having two places to specify the same thing feel weird
- introduced by @hubertp in #3791

> Builtins...carry a `autoRegister` property to determine if a builtin method should be automatically registered with the type.

What's the benefit of _registering a builtin automatically with the type_, @hubertp?

hubertp replied:
It's been a while ago but IIRC that was specifically requested either by you or @kustosz, so that we don't have to write all those builtin methods as it is done now in this PR.

JaroslavTulach reviewed (changes requested):
I don't understand the (amount of) `@Builtin_Method` changes.

Akirathan commented on engine/common/src/main/java/org/enso/common/CompilationStage.java:
```diff
@@ -3,12 +3,13 @@
 /** Defines a stage of compilation of the module. */
 public enum CompilationStage {
   INITIAL(0),
-  AFTER_PARSING(1),
```
What is the purpose of this change?

radeusgd replied:
I was introducing a new stage: `AFTER_TYPE_INFERENCE_PASSES`, that was supposed to be between two other existing passes. I thought that multiplying the numbers by 10 allows us to introduce such middle stages in the future in an easier manner - one would not need to modify all values as there will still be 'free' integers between existing stage numbers.

But I can just bring this back to a natural 0, 1, 2... ordering if that's preferred - please let me know.

Akirathan commented on engine/runtime/src/main/java/org/enso/interpreter/runtime/scope/ModuleScope.java:
```diff
@@ -20,14 +22,16 @@
   private final Map<Type, Map<String, Supplier<Function>>> methods;
 
   /**
+   * First key is target type, second key is source type. The value is the conversion function from
```
Duplicated line

Akirathan commented on engine/runtime-integration-tests/src/test/java/org/enso/compiler/test/StaticAnalysisTest.java:
```diff
@@ -0,0 +1,138 @@
+          .invokeMember(MethodNames.TopScope.LEAK_CONTEXT)
+          .asHostObject();
+
+  protected Module compile(Source src) {
```
DRY: I believe that instead of this, and manual context creation, you could use methods from `ContextUtils`, like https://github.com/enso-org/enso/blob/fb821303ab2818c4ce2d8482612584497f2e2014/lib/java/test-utils/src/main/java/org/enso/test/utils/ContextUtils.java#L183

radeusgd replied:
Right. I think that these just did not exist when I wrote this, and I did not notice now that this was introduced - thanks for letting me know!

radeusgd replied:
After further inspection, my `compile` is doing more than the `ContextUtils::compile` - it is also registering the modules in scope so that they can be imported - which is crucial for tests that test behaviour of type inference across different modules. I guess perhaps I can try porting my additional 'features' to `ContextUtils`.

radeusgd replied:
I have kept my `compile` separate because IMO it differs more than is similar, because of the additional module management, so in my opinion it is not really a DRY problem - the repetition is very small and more is changed than not.

@Akirathan If you still think it should be merged then please let me know, we will find a way.

Akirathan commented on engine/runtime-compiler/src/main/java/org/enso/compiler/pass/analyse/types/scope/TypeHierarchy.java:
```diff
@@ -0,0 +1,41 @@
+ * This should be aligned with Type.allTypes in the interpreter.
+ */
+public class TypeHierarchy {
+  public TypeScopeReference getParent(TypeScopeReference type) {
```
Refactor to static method? No need for this to be instance.

Akirathan commented on engine/runtime-compiler/src/main/java/org/enso/compiler/pass/analyse/types/BuiltinTypes.java:
```diff
@@ -5,15 +5,11 @@
 
 /** A helper class providing the builtin types. */
 public class BuiltinTypes {
```
Why are not all the methods in this class static, and this class does not have a private constructor?

radeusgd replied:
Good question. Initially I thought that I will need to initialize the `BuiltinTypes` with some references to the compiler context to be able to know more info about these builtins. After all it was handled a different way. I did not change it as I did not want to close the way for `BuitlinTypes` having some members that would have to be initialized.

But given that for now it _could_ be static, I guess I can change it, and if the need ever arises, revert such change. As you are right that for now this does not serve much purpose.

Akirathan commented on engine/runtime-integration-tests/src/test/java/org/enso/compiler/test/TypesFromSignaturesTest.java:
```diff
@@ -0,0 +1,186 @@
+
+  private void assertInferredType(IR ir, String expected) {
+    TypeRepresentation inferred = getInferredType(ir);
+    assertEquals(expected, inferred.toString());
```
Minor: Relying on correct `toString` implementation in tests does not seem like a good idea. What about moving the current `toString` implementation to a class in tests dir in the same package? It would then look something like this:
```java
assertEquals(expected, TestUtil.toString(inferred));
```

radeusgd replied:
> Relying on correct `toString` implementation in tests does not seem like a good idea.

Hmm, I never looked at it this way. IMHO it may make sense to test `toString`, but I guess it's a habit from working on Libs where the `to_text` implementation is part of the public user-facing API and is _supposed_ to be well tested.

I'm happy to move it to a helper if you think that will be better.

JaroslavTulach replied:
> Relying on correct `toString` implementation in tests does not seem like a good idea

Why? `toString()` format is part of an API. But yeah, having for example:
```
assertTypeRepresentation(....)
```
might be more flexible.

Akirathan commented on engine/runtime-integration-tests/src/test/java/org/enso/compiler/test/StaticAnalysisTest.java:
```diff
@@ -0,0 +1,138 @@
+import org.graalvm.polyglot.Source;
+import scala.jdk.javaapi.CollectionConverters;
+
+public abstract class StaticAnalysisTest {
```
Methods in this class seems useful even outside of static analysis. Could you try either merging them to [ContextUtils](https://github.com/enso-org/enso/blob/fb821303ab2818c4ce2d8482612584497f2e2014/lib/java/test-utils/src/main/java/org/enso/test/utils/ContextUtils.java) or just to the same package and refactoring the methods here as static?

radeusgd replied:
I will see what I can do.

This PR was in progress for such a long time (because I was rarely finding time to work on it), that I think `StaticAnalysisTest` was created before `ContextUtils`. I guess if I were creating it now I'd indeed put them somewhere around there.

Akirathan reviewed (commented):
Seems good overall. Published few minor comments. How am I supposed to actually run the static type checker from the command line? I have tried
```
enso --enable-static-analysis --no-ir-caches --compile ~/tmp/Proj
```
on a project with this single source `Main.enso`:
```
from Standard.Base import all

foo x:Integer -> Integer =
    x + 1

main =
    foo "Hello"
```
and expected an error, but it normally compiles.

Moreover, do we have an issue that tracks integration of static type checking into our tests?

Akirathan commented:
Could you also add another test that creates a dummy project and invokes the type checker and checks that there are some warnings? Something like this:
```diff
diff --git a/engine/runtime-integration-tests/src/test/java/org/enso/compiler/test/TypeInferenceTest.java b/engine/runtime-integration-tests/src/test/java/org/enso/compiler/test/TypeInferenceTest.java
index 416c067a0..c4999481f 100644
--- a/engine/runtime-integration-tests/src/test/java/org/enso/compiler/test/TypeInferenceTest.java
+++ b/engine/runtime-integration-tests/src/test/java/org/enso/compiler/test/TypeInferenceTest.java
@@ -2,14 +2,20 @@ package org.enso.compiler.test;
 
 import static org.junit.Assert.assertEquals;
 import static org.junit.Assert.assertNotEquals;
+import static org.junit.Assert.assertNotNull;
 import static org.junit.Assert.assertTrue;
 import static org.junit.Assert.fail;
 
+import java.io.ByteArrayOutputStream;
+import java.io.IOException;
 import java.net.URI;
+import java.nio.file.Files;
+import java.nio.file.Path;
 import java.util.HashSet;
 import java.util.List;
 import java.util.Optional;
 import java.util.Set;
+import org.enso.common.RuntimeOptions;
 import org.enso.compiler.core.IR;
 import org.enso.compiler.core.ir.Module;
 import org.enso.compiler.core.ir.ProcessingPass;
@@ -19,11 +25,19 @@ import org.enso.compiler.core.ir.module.scope.definition.Method;
 import org.enso.compiler.pass.analyse.types.InferredType;
 import org.enso.compiler.pass.analyse.types.TypeInferencePropagation;
 import org.enso.compiler.pass.analyse.types.TypeRepresentation;
+import org.enso.test.utils.ContextUtils;
+import org.enso.test.utils.ProjectUtils;
 import org.graalvm.polyglot.Source;
 import org.junit.Ignore;
 import org.junit.Test;
 import scala.Option;
 
+import static org.hamcrest.MatcherAssert.assertThat;
+import static org.hamcrest.Matchers.is;
+import static org.hamcrest.Matchers.notNullValue;
+import static org.hamcrest.Matchers.containsString;
+
+
 public class TypeInferenceTest extends StaticAnalysisTest {
   @Test
   public void zeroAryCheck() throws Exception {
@@ -1430,6 +1444,38 @@ public class TypeInferenceTest extends StaticAnalysisTest {
     assertAtomType("local.Project1.modB.Typ_Y", findAssignment(foo, "x2"));
   }
 
+  @Test
+  public void staticTypeCheckerReportsWarningsOnProject() throws IOException {
+    var mainSrc = """
+        type My_Type
+            Cons data
+        
+        foo x:My_Type -> My_Type =
+            My_Type.Cons (x.data + 1)
+        
+        bar =
+            foo "Hello"
+        
+        main =
+            42
+        """;
+    Path projDir = Files.createTempDirectory("enso-tests");
+    ProjectUtils.createProject("Proj", mainSrc, projDir);
+    var out = new ByteArrayOutputStream();
+    var ctxBuilder= ContextUtils.defaultContextBuilder()
+        .option(RuntimeOptions.DISABLE_IR_CACHES, "true")
+        .option(RuntimeOptions.ENABLE_STATIC_ANALYSIS, "true")
+        .option(RuntimeOptions.STRICT_ERRORS, "true")
+        .out(out)
+        .err(out)
+        .logHandler(out);
+    ProjectUtils.testProjectRun(ctxBuilder, projDir, res -> {
+      assertThat(res.isNumber(), is(true));
+      assertThat(res.asInt(), is(42));
+      assertThat(out.toString(), containsString("Type mismatch"));
+    });
+  }
+
   private TypeRepresentation getInferredType(IR ir) {
     var option = getInferredTypeOption(ir);
     assertTrue(
```

JaroslavTulach commented on engine/runtime-compiler/src/main/java/module-info.java:
```diff
@@ -29,4 +29,5 @@
   exports org.enso.compiler.phase;
   exports org.enso.compiler.phase.exports;
   exports org.enso.compiler.refactoring;
+  exports org.enso.compiler.common_logic;
```
No snakecase, please. That's not typical Java convention.

JaroslavTulach commented on engine/runtime-compiler/src/main/java/org/enso/compiler/common_logic/BuildScopeFromModuleAlgorithm.java:
```diff
@@ -0,0 +1,191 @@
+  }
+
+  /** Runs the main processing on a module, that will build the module scope for it. */
+  public void processModule(Module moduleIr, BindingsMap bindingsMap) {
```
Shall this method be overriden? If not, make it `final`.

JaroslavTulach commented on engine/runtime-compiler/src/main/java/org/enso/compiler/common_logic/BuildScopeFromModuleAlgorithm.java:
```diff
@@ -0,0 +1,191 @@
+  private final Logger logger = LoggerFactory.getLogger(BuildScopeFromModuleAlgorithm.class);
+
+  /** The scope builder to which the algorithm will register the entities. */
+  protected final ModuleScopeBuilderType scopeBuilder;
```
Why is this field `protected`?

radeusgd replied:
I think child classes were also using it in the implementations of the `protected` methods. If that ceased to be the case and I forgot to notice, I will make it `private`. Otherwise, if it's used in child classes, do you think it can stay `protected`?

Because alternatively, I could make it `private` and store yet another reference to the scope builder in each child class. But storing the same reference twice in the same class felt weird for me and I didn't think it was necessary, that's why I opted for `protected final`. But I'm open to a different approach if you think it's preferred.

JaroslavTulach commented on engine/runtime-compiler/src/main/java/org/enso/compiler/common_logic/BuildScopeFromModuleAlgorithm.java:
```diff
@@ -0,0 +1,191 @@
+    TypeScopeReferenceType,
+    ImportExportScopeType,
+    ModuleScopeType extends
+        CommonModuleScopeShape<FunctionType, TypeScopeReferenceType, ImportExportScopeType>,
```
Wow! One should always beware of what one asks for!

> ModuleScopeType extends
>        CommonModuleScopeShape<FunctionType, TypeScopeReferenceType, ImportExportScopeType

Showing [object algebras](http://wiki.apidesign.org/wiki/ObjectAlgebras) to a functional programmers may lead to code that I no longer understand...

JaroslavTulach replied:
I don't think there is any need for `T1 extends CT1<....>`.

If you want to perform operation on `ModuleScopeType`, then add a `protected abstract` method that works on `ModuleScopeType`.

Avoid `extends` in the algorithm generic arguments.

radeusgd replied:
Hmmm, okay I can rewrite it to such a form. It felt to me that it's a bit easier and allows me to separate concerns - the parts providing the `ModuleScopeType` interface were separated from the main algorithm and could be shared across 2 different 'algorithm' abstract classes.

If I put these methods directly into the algorithm, I'll need to duplicate them for each one. How can I ensure I can still share them? I guess I need to create a separate class `ModuleScopeInterface<ModuleScopeType>` that will allow me to act on the `ModuleScopeType` values?

JaroslavTulach replied:
> Hmmm, okay I can rewrite it to such a form.

That's indeed my preference. As an alternative...

> need to duplicate them for each one

... you can try to convince me that your design is right. I just feel scared for a signature like:
```
ModuleScopeType extends
        CommonModuleScopeShape<FunctionType, TypeScopeReferenceType, ImportExportScopeType>
```

radeusgd replied:
In these three commits I have 'simplified' the algorithms by removing the `extends` and replacing the interfaces with additional abstract methods: https://github.com/enso-org/enso/pull/11399/files/6b920eb565c878c136a864604b3557c5b1fb2623..dfa200b1d13dde772c688deeb116df46d9ee2c1b

Personally, I don't like it. It adds more boilerplate (the additional abstract methods) and the removal of `CommonModuleScopeShape` severs the connection between `ModuleScope` and `StaticModuleScope` (and their builder respectively). That was added deliberately as a kind of 'scaffolding' that ensures that as much as possible of common structure is shared between these two related-but-apart classes - as a way to try to ensure that any changes to one will be reflected in the other. Now I feel like the connection is a bit more loose, less rigid, and changes to structure of `ModuleScope` may be forgotten to be reflected in `StaticModuleScope`. Of course the `CommonModuleScopeShape` was not a perfect synchronization and it was possible still to make incompatible changes, but it was a way to relate the two classes.

Anyway, if you think that the code is better that way, I'm ok to keep it.

JaroslavTulach commented on engine/runtime-compiler/src/main/java/org/enso/compiler/common_logic/MethodResolutionAlgorithm.java:
```diff
@@ -0,0 +1,212 @@
+   *   <li>Finally, methods imported from other modules.
+   * </ol>
+   */
+  public FunctionType lookupMethodDefinition(
```
To be overriden or not? If not, make `final`.

JaroslavTulach reviewed (changes requested):
- avoid snake case in Java code
- make `public` methods `final` when they aren't supposed to be overriden
- avoid `extends` in algorithms - I believe they aren't necessary and just complicate the algorithm with external references.

hubertp commented on engine/runtime/src/main/scala/org/enso/interpreter/runtime/IrToTruffle.scala:
```diff
@@ -185,80 +182,215 @@ class IrToTruffle(
+    override protected def processConversion(
+      conversion: Method.Conversion
+    ): Unit = {
+      lazy val where =
```
Change it do `def` and make `rootScopeInfo` accept a function instead. Having a lazy val here and below is rather pointless. I know it has been like that before :)

radeusgd replied:
Hmmmm 🤔 Like this? 179a816a

Tbh I prefer the original approach with lazy val and by-name passing, it feels clearer to me. But if you prefer the `def` I'm happy to keep it.

hubertp commented on engine/runtime/src/main/scala/org/enso/interpreter/runtime/IrToTruffle.scala:
```diff
@@ -185,80 +182,215 @@ class IrToTruffle(
+            )
+          case _ =>
+            throw new CompilerError(
+              "Conversion bodies must be functions at the point of codegen."
```
include `toType` and `fromType` for easier identification of bug

hubertp commented on engine/runtime/src/main/scala/org/enso/interpreter/runtime/IrToTruffle.scala:
```diff
@@ -185,80 +182,215 @@ class IrToTruffle(
+    override protected def processMethodDefinition(
+      method: Method.Explicit
+    ): Unit = {
+      lazy val where =
```
same comment as above

hubertp commented on lib/scala/pkg/src/main/scala/org/enso/pkg/QualifiedName.scala:
```diff
@@ -67,6 +67,16 @@ object QualifiedName {
+        "Cannot create a qualified name from an empty list of parts."
+      )
+    }
+    QualifiedName(parts.dropRight(1), parts.last)
```
```suggestion
    QualifiedName(parts.init, parts.last)
```

hubertp reviewed (commented):
There is a lot of changes here, making it hard to review. It would help, in the future, if scope changes were committed separately because I think they can.

I don't see the need for explicit builtin definitions so maybe I'm missing something. Auto-register was defined for that. If no longer allow/desire it then that functionality can go as well.

radeusgd commented:
> There is a lot of changes here, making it hard to review. It would help, in the future, if scope changes were committed separately because I think they can.

They certainly couldn't have been done separately, as I had to first figure out what I need in `StaticModuleScope` to be able to find what is _common enough_ that can be relatively easily extracted to the common base. Now, having created it, yes, I could artificially split the code changes into 2 separate patches to make the review easier - it just required additional work that I thought maybe will not be necessary. But please let me know if I should do that. Just noting that I could not have avoided the additional work of splitting by 'better planning', as at the beginning I didn't know what the split would have to look like.

radeusgd committed:
Merge branch 'develop' into wip/radeusgd/9812-infer-method-calls

# Conflicts:
#	distribution/lib/Standard/Base/0.0.0-dev/src/Meta/Enso_Project.enso
#	engine/runtime-integration-tests/src/test/java/org/enso/compiler/test/TypeInferenceTest.java
#	engine/runtime/src/main/java/org/enso/interpreter/runtime/scope/ModuleScope.java
#	engine/runtime/src/main/scala/org/enso/interpreter/runtime/IrToTruffle.scala

radeusgd committed:
fixes after merge to make it compile

radeusgd committed:
Merge branch 'develop' into wip/radeusgd/9812-infer-method-calls

radeusgd committed:
fix tests - modules with qualified name need a Package

radeusgd committed:
Merge branch 'develop' into wip/radeusgd/9812-infer-method-calls

radeusgd committed:
CR make BuiltinTypes static, move helpers

radeusgd committed:
CR remove unused method

radeusgd committed:
CR replace lazy val with def?

radeusgd committed:
CR rename package avoiding snake case

radeusgd committed:
fmt

radeusgd committed:
CR remove outdated builtin shadow definition

radeusgd committed:
CR: simplify the algorithm abstractions

radeusgd committed:
CR: simplify the algorithm abstractions - pt. 2

radeusgd committed:
further cleanup after simplifications

radeusgd committed:
CR make some functions final

radeusgd committed:
Merge branch 'develop' into wip/radeusgd/9812-infer-method-calls

radeusgd committed:
CR refactor to static

radeusgd commented:
I think I tried addressing all of the comments. I will very much appreciate your reviews on this next iteration of the PR. Sorry it is happening after such a long time - I will try to schedule time to iterate on this PR faster this time.

radeusgd committed:
avoid NoSuchElementException in assert

radeusgd committed:
try fixing failing asserts `validateConsistency`;

@Akirathan suggested to remove his `validateConsistency` check, but I've found I could fix the asserts by caching the returned value in persisted Map (if there was no caching, then on every ir.passData().get(...) a new instance of metadata was recreated, so the `validateConsistency` was failing because it requires object _identity_ and not only equality).

We can either keep this 'cache' or remove Pavel's check.

radeusgd committed:
Revert "try fixing failing asserts `validateConsistency`;"

This reverts commit 9d5deaa370030ef97a2545711f1e3cc3080738bc.

radeusgd committed:
remove assert as requested by @Akirathan

radeusgd committed:
Merge branch 'develop' into wip/radeusgd/9812-infer-method-calls

# Conflicts:
#	engine/runtime/src/main/scala/org/enso/interpreter/runtime/IrToTruffle.scala

radeusgd commented:
I'm getting weird test failures in `ModuleManagementTest`, removing the assertion as @Akirathan suggested has changed the error but it is still failing at deserialization in a weird way:
```
[info] - should allow importing literal-source modules *** FAILED *** (1 second, 195 milliseconds)
[info]   org.graalvm.polyglot.PolyglotException: java.lang.ArrayIndexOutOfBoundsException: Index 158052 out of bounds for length 1
[info]   at org.enso.persistance/org.enso.persist.PerInputImpl$InputCache.getRef(PerInputImpl.java:351)
[info]   at org.enso.persistance/org.enso.persist.PerInputImpl.findReference(PerInputImpl.java:58)
[info]   at org.enso.persistance/org.enso.persist.PerReferencePeristance.readObject(PerReferencePersistance.java:33)
[info]   at org.enso.persistance/org.enso.persist.PerReferencePeristance.readObject(PerReferencePersistance.java:8)
[info]   at org.enso.persistance/org.enso.persist.Persistance.readWith(Persistance.java:162)
[info]   at org.enso.persistance/org.enso.persist.PerInputImpl.readInline(PerInputImpl.java:67)
[info]   at org.enso.runtime/org.enso.interpreter.caches.PersistBindingsMap_Argument.readObject(PersistBindingsMap_Argument.java:14)
[info]   at org.enso.runtime/org.enso.interpreter.caches.PersistBindingsMap_Argument.readObject(PersistBindingsMap_Argument.java:4)
[info]   at org.enso.persistance/org.enso.persist.Persistance.readWith(Persistance.java:162)
[info]   at org.enso.persistance/org.enso.persist.PerInputImpl.readIndirect(PerInputImpl.java:218)
[info]   ...
```

Interestingly, it seems that my followup PR #11955 that removes the `BindingsMap -> Expression` cycle from the IR seems to not have these failures.

@JaroslavTulach do you think we can merge both PRs at once (I need to finish the followup but I'm working on that) OR would you like to help me investigating this serialization issue? Seems like there is still some issue in the persistance when cycles are around.

IMO it will be easier to just merge the followup, although we can also try investigating the persistance to make it more robust as such failures probably shouldn't be happening - but I don't think fixing persistance should be the focus of this PR in more extent than necesary.

radeusgd committed:
add example test from @Akirathan but adapted to this PR's capabilities (no parameter type mismatch checks yet, instead testing method mismatch)

radeusgd commented:
I wanted to keep these PRs separate but it initially (I'm waiting for full CI suite to pass to verify) seems that the followup PR #11955 is fixing the persistance issues that are causing test failures in this PR. So I think ideally I'd merge #11955 into this PR and then merge the whole changeset as a single change. I'm not merging yet so that you can review more easily - first my answers to your requests in this PR and then the followup PR that builds on top of it.

JaroslavTulach commented on engine/common/src/main/java/org/enso/common/CompilationStage.java:
```diff
@@ -3,12 +3,13 @@
+  AFTER_GLOBAL_TYPES(30),
+  AFTER_STATIC_PASSES(40),
+  AFTER_TYPE_INFERENCE_PASSES(45),
+  AFTER_RUNTIME_STUBS(50),
```
Can we finish `AFTER_RUNTIME_STUBS` while not performing `AFTER_TYPE_INFERENCE_PASSES`? I hope we can as the _type inference_ was supposed to be optional.

Then I am afraid this _linear order_ may no longer be sufficient. We may need a _bit mask_...

radeusgd replied:
We can move to a bit mask but I think it would make things more complicated the necessary.

Currently the logic is following - `AFTER_TYPE_INFERENCE_PASSES` is set after `runFinalTypeInferencePasses` has been run. It is in fact run **always**, **but** the pass group that it runs (`typeInferenceFinalPasses`) is empty if the static analysis is not explicitly enabled. So by default it just runs an 'empty' run - which basically should do nothing.

Thus we can still keep a linear order of passes, while having the static analysis be optional to run.

Is that acceptable?

JaroslavTulach commented on engine/runtime-compiler/src/main/java/org/enso/compiler/common/BuildScopeFromModuleAlgorithm.java:
```diff
@@ -0,0 +1,180 @@
+   *   <li>For a module method, this will be the type associated with the module.
+   * </ul>
+   */
+  protected final TypeScopeReferenceType getTypeAssociatedWithMethod(Method.Explicit method) {
```
In Java reflection API this is called _method defining type_...

JaroslavTulach commented on engine/runtime-compiler/src/main/java/org/enso/compiler/common/BuildScopeFromModuleAlgorithm.java:
```diff
@@ -0,0 +1,180 @@
+
+  protected abstract void registerExport(ImportExportScopeType exportScope);
+
+  protected abstract void registerImport(ImportExportScopeType importScope);
```
I assume these `protected abstract void` methods encourage a non-functional programming style where the state of the _algorithm_ is being modified while the algorithm is still computing something. A functional algorithm would collect all such information itself and would expose it for further processing once its `processModule` method is over.

radeusgd replied:
This is aligned with how the pre-existing `ModuleScope.Builder` used to work. I wrote the `Static` counterpart in the same shape - that was my idea from the beginning - for the static analysis to reflect the shape of its 'dynamic' counterpart.

Refactoring the original algorithm into more functional style can be done, but I think it is out of scope of the type changes. It would introduce even more changes, making this even harder to review than currently (and there were already comments that this change is too large).

JaroslavTulach commented on engine/runtime-compiler/src/main/java/org/enso/compiler/pass/analyse/types/BuiltinTypes.java:
```diff
@@ -4,31 +4,47 @@
 
-  public final TypeRepresentation INTEGER = fromQualifiedName("Standard.Base.Data.Numbers.Integer");
-  public final TypeRepresentation FLOAT = fromQualifiedName("Standard.Base.Data.Numbers.Float");
+  public static final String numberQualifiedName = "Standard.Base.Data.Numbers.Number";
```
Java naming convention for string constants suggests to use `NUMBER_QUALIFIED_NAME` - btw. `FQN_NUMBER` might be shorter all such constants would have a short common prefix.

JaroslavTulach commented on engine/runtime-compiler/src/main/java/org/enso/compiler/pass/analyse/types/TypeInferencePropagation.java:
```diff
@@ -66,32 +70,54 @@
-  private final BuiltinTypes builtinTypes = new BuiltinTypes();
+public final class TypeInferencePropagation implements IRPass {
+  public static final TypeInferencePropagation INSTANCE = new TypeInferencePropagation();
+  private static final Logger logger = LoggerFactory.getLogger(TypeInferencePropagation.class);
```
Java naming convention would suggest `LOGGER`.

JaroslavTulach commented on engine/runtime-compiler/src/main/java/org/enso/compiler/pass/analyse/types/TypeInferenceSignatures.java:
```diff
@@ -0,0 +1,195 @@
+ * org.enso.compiler.core.TreeToIr}). It does not look any deeper into the expressions, ensuring
+ * that it is relatively quick to run.
+ */
+public class TypeInferenceSignatures implements IRPass {
```
```suggestion
public final class TypeInferenceSignatures implements IRPass {
```

and also `private` constructor.

JaroslavTulach commented on engine/runtime-compiler/src/main/java/org/enso/compiler/pass/analyse/types/TypeResolver.java:
```diff
@@ -24,43 +26,10 @@
-        }
-      }
+      case Name.Literal name -> getResolvedTypeFromBindingsMap(name);
+      case Name.SelfType selfType -> getResolvedTypeFromBindingsMap(selfType);
```
Yup `Self` is supported.

JaroslavTulach commented on engine/runtime-compiler/src/main/java/org/enso/compiler/pass/analyse/types/scope/BuiltinsFallbackScope.java:
```diff
@@ -0,0 +1,37 @@
+/**
+ * This is a special scope that notes methods which are always available on Any type.
+ *
+ * <p>They are available even without any imports.
```
- see #8852 for some thoughts on what should be available without any import

JaroslavTulach commented on engine/runtime-compiler/src/main/java/org/enso/compiler/pass/analyse/types/scope/BuiltinsFallbackScope.java:
```diff
@@ -0,0 +1,37 @@
+ * <p>They are available even without any imports.
+ */
+public class BuiltinsFallbackScope {
+  private StaticModuleScope cachedAnyScope = null;
```
Given there is just a single method in this class:
- when the class is loaded into (HotSpot) JVM
- it is because `fallbackAnyScope()` is about to be called
- e.g. you can initialize `cachedAnyScope` in static initializer and make it `final`
- and then just return it
- it will remain as lazy as this code is

radeusgd replied:
I have to admit that I thought that the static initializers are all executed when the JARs are loaded, eagerly, not lazily. [TIL](https://docs.oracle.com/javase/specs/jls/se8/html/jls-12.html#jls-12.4).

If it's lazy, I can indeed simplify the code as suggested.

JaroslavTulach commented on engine/runtime-compiler/src/main/java/org/enso/compiler/pass/analyse/types/scope/ModuleResolver.java:
```diff
@@ -0,0 +1,37 @@
+import org.slf4j.LoggerFactory;
+
+/** A helper class that allows to resolve qualified names to loaded modules. */
+public class ModuleResolver {
```
```suggestion
public final class ModuleResolver {
```

JaroslavTulach commented on engine/runtime-compiler/src/main/java/org/enso/compiler/pass/analyse/types/scope/StaticImportExportScope.java:
```diff
@@ -0,0 +1,83 @@
+import org.enso.pkg.QualifiedName;
+
+/** The static counterpart of {@link org.enso.interpreter.runtime.scope.ImportExportScope}. */
+public class StaticImportExportScope {
```
```suggestion
public final class StaticImportExportScope {
```

JaroslavTulach commented on engine/runtime-compiler/src/main/java/org/enso/compiler/pass/analyse/types/scope/StaticImportExportScope.java:
```diff
@@ -0,0 +1,83 @@
+    this.referredModuleName = referredModuleName;
+  }
+
+  private transient MaterializedImportExportScope cachedMaterializedScope = null;
```
`transient`!? Such a keyword only has a meaning in `java.io.Serializable` classes.

radeusgd replied:
I tend to use this as a kind of 'comment' that the value is just a cache that can be set to null without any issue.

But perhaps it is a bit of abuse of the keyword. I can remove it.

JaroslavTulach commented on engine/runtime-compiler/src/main/java/org/enso/compiler/pass/analyse/types/scope/StaticImportExportScope.java:
```diff
@@ -0,0 +1,83 @@
+    return materialized;
+  }
+
+  public static class MaterializedImportExportScope {
```
What is the purpose of "materialized"? When things get materialized? What does it mean they are "materialized"?

radeusgd replied:
This is similar to the abstract vs concrete distinction in `ResolvedName` - it ensures we don't store references to other modules in the metadata as that is problematic with serialization (a reference to Standard.Base module from Table would essentially mean when serializing Table, a lot of stuff from Base would end up there as well).

But instead of using `ResolvedName` and a complex reconstruction process in `prepareForSerialization`/`restoreFromSerialization`, I just store the `QualifiedName` and resolve it on-demand. That resolution is 'materialization'. So the materialized counterpart is never stored in the metadata, it is only constructed ad-hoc during processing (that's also why it's cached).

radeusgd replied:
I guess this is a bit redundant to `ResolvedName`. I could try to convert it to use `ResolvedName` instead, although it may be quite hard to do so. The problem is that also the `TypeRepresentation` relies on just `QualifiedName` instead of `ResolvedName`. Bringing `ResolvedName` into `TypeRepresentation` would make it no longer a simple structure but a structure that needs complex preprocessing before serialization. And I think the added complexity of that is just not worth it - I don't see much benefits from it.

radeusgd replied:
Even if we decide to do so, I'd suggest to do it in a separate PR as it would be a relatively large change again.

JaroslavTulach replied:
Materialization has a bit different meaning in Truffle - see [MaterializedFrame](https://www.graalvm.org/truffle/javadoc/com/oracle/truffle/api/frame/MaterializedFrame.html).

> That resolution is 'materialization'

- call it `resolve`
- rename `MaterializedImportExportScope` to just `Resolved`
- thus forming `StaticImportExportScope.Resolved` full (non qualified) name

maybe?

radeusgd replied:
I see, now I understand why the name was misleading. Thanks for the suggestion how I can rename it :)

JaroslavTulach commented on engine/runtime-compiler/src/main/java/org/enso/compiler/pass/analyse/types/scope/StaticMethodResolution.java:
```diff
@@ -0,0 +1,123 @@
+        TypeRepresentation, TypeScopeReference, StaticImportExportScope, StaticModuleScope> {
+  private final ModuleResolver moduleResolver;
+  private final BuiltinsFallbackScope builtinsFallbackScope;
+  private final Logger logger = LoggerFactory.getLogger(this.getClass());
```
```suggestion
  private static final Logger logger = LoggerFactory.getLogger(StaticMethodResolution.class);
```
also Java guidelines would suggest to rename to `LOGGER`.

radeusgd replied:
Thanks for this and the related suggestions, I somehow missed that. I will rename them.

JaroslavTulach commented on engine/runtime-compiler/src/main/java/org/enso/compiler/pass/analyse/types/scope/StaticModuleScope.java:
```diff
@@ -0,0 +1,157 @@
+ * This is a sibling to the ModuleScope.
+ *
+ * <p>The ModuleScope is the runtime representation of a module, optimized for fast runtime
+ * dispatch. The StaticModuleScope is an analogous structure, that can be used by static analysis
```
`ModuleScope` isn't optimized for _fast runtime dispatch_ that much. It is rather a _huge hash map_ to look things up. Such a lookup is usually hidden behind `@TrufleBoundary` - e.g. it is slow. The fast dispatch is then implemented elsewhere in Enso Truffle `Node` subclasses.

JaroslavTulach commented on engine/runtime-compiler/src/main/java/org/enso/compiler/pass/analyse/types/scope/StaticModuleScope.java:
```diff
@@ -0,0 +1,157 @@
+ * with Types and their Constructors that are used during static resolution of some names. This
+ * class also keeps track of all defined methods, to facilitate type checking. I'm keeping these
+ * separate for now as it is easier to create a prototype that way. If later we find out they have
+ * enough of similarity, we should merge them.
```
I'd like to see simplification/deprecation/removal of `BindingsMap` with an object that is _immutable_ as much as possible. `BindingsMap` mutates its state (during import/export resolution) and it is causing headache (it least to me).

radeusgd replied:
Perhaps we could move toward using `StaticBindingsMap` then. For now it is only constructed if the static analysis is enabled (as I haven't yet measured the performance impact). But once we use it more, maybe it could replace `BindingsMap` altogether.

JaroslavTulach commented on engine/runtime-compiler/src/main/java/org/enso/compiler/pass/analyse/types/scope/StaticModuleScope.java:
```diff
@@ -0,0 +1,157 @@
+
+    public StaticModuleScope build() {
+      sealed = true;
+      return new StaticModuleScope(
```
This looks like _immutable enough_ structure:
- it has `Builder`
- the builder can only be used once and then it gets sealed
- collections are then immutable

JaroslavTulach commented on engine/runtime-compiler/src/main/java/org/enso/compiler/pass/analyse/types/scope/StaticModuleScope.java:
```diff
@@ -0,0 +1,157 @@
+    return associatedType;
+  }
+
+  public static StaticModuleScope forIR(Module module) {
```
I see this as a nice entry point for obtaining `StaticModuleScope` for an `ir.Module`. Maybe it deserves more Javadoc to:
- describe what pass needs to run before this method is used
- what happens, if this method is called before/without the pass being performed

JaroslavTulach commented on engine/runtime-compiler/src/main/java/org/enso/compiler/pass/analyse/types/scope/StaticModuleScopeAnalysis.java:
```diff
@@ -0,0 +1,201 @@
+      StaticModuleScope.Builder scope,
+      TypeScopeReference typeScope,
+      Definition.Type typeDefinition) {
+    HashMap<String, List<TypeRepresentation>> fieldTypes = new HashMap<>();
```
```suggestion
    var fieldTypes = new HashMap<String, List<TypeRepresentation>>();
```

radeusgd replied:
why is `var` preferred here?

Personally I prefer how it is now - because the type of the mapping (key and value type) is more clear IMO.

JaroslavTulach replied:
I'd be more Scala-like ;-)

radeusgd replied:
> I'd be more Scala-like ;-)

I'd disagree, in Scala you can still add type to the local definitions, e.g.:
```
val fieldTypes: Map[String, List[Typ]] = Map.empty
```
and I think it is useful to add it when the type serves as a comment clarifying what the mapping is holding.

JaroslavTulach replied:
Now I realized that I don't like `HashMap` at the beginning. It should be:
```java
Map<String, List<TypeRepresentation>> fieldTypes = new HashMap<>();
```
e.g. [code against interfaces and not implementation](http://wiki.apidesign.org/wiki/Code_Against_Interfaces%2C_Not_Implementations). Contrary to such a _wish_ of mine, the
```java
var fieldTypes = new HashMap<String, List<TypeRepresentation>>();
```
does the opposite - e.g. it exposes `HashMap` as a type of `fieldTypes` - but it is not that visible...

radeusgd replied:
Yeah `Map` looks best there, I will update to that.

radeusgd replied:
Updated in [22e6649](https://github.com/enso-org/enso/pull/11399/commits/22e66494f19abf38cea619ed7e521bc6280670d4)

JaroslavTulach commented on engine/runtime-compiler/src/main/java/org/enso/compiler/pass/analyse/types/scope/StaticModuleScopeAnalysis.java:
```diff
@@ -0,0 +1,201 @@
+      Definition.Type typeDefinition) {
+    HashMap<String, List<TypeRepresentation>> fieldTypes = new HashMap<>();
+    for (var constructorDef : CollectionConverters$.MODULE$.asJava(typeDefinition.members())) {
+      for (var argumentDef : CollectionConverters$.MODULE$.asJava(constructorDef.arguments())) {
```
Consider using https://github.com/enso-org/enso/blob/b5f110617e1b96c118c6483bddaf982c484c8945/lib/java/scala-libs-wrapper/src/main/java/org/enso/scala/wrapper/ScalaConversions.java

JaroslavTulach commented on engine/runtime-compiler/src/main/java/org/enso/compiler/pass/analyse/types/BuiltinTypes.java:
```diff
@@ -4,31 +4,47 @@
   public static boolean isAny(QualifiedName qualifiedName) {
     var str = qualifiedName.toString();
-    return str.equals("Standard.Base.Any.Any") || str.equals("Standard.Base.Any");
+    return str.equals(anyQualifiedName) || str.equals("Standard.Base.Any");
```
Two names for `Any`? Isn't the first one the `Any` type and the second one the `Any.enso` module?

radeusgd replied:
Thanks for making me rethink that!

It used to be a hack that made something work. But it seems that it is no longer needed, so indeed it should have been removed.

JaroslavTulach commented on engine/runtime-compiler/src/main/scala/org/enso/compiler/Compiler.scala:
```diff
@@ -133,12 +133,17 @@ class Compiler(
     */
   def compile(
     shouldCompileDependencies: Boolean,
+    shouldWriteCache: Boolean,
```
I assume this is a reasonable way to avoid the _static type information_ to be stored in the caches.

radeusgd replied:
Yes, plus IIRC it fixes a weird behaviour - that even if caches were disabled, they would still be written after compilation or something.

Now, if we run `--no-ir-cache`, we neither _read_ **nor _write_** the caches, which I think is correct.

JaroslavTulach replied:
The reason why we are seeing failures like https://github.com/enso-org/enso/actions/runs/12680614808/job/35342759884?pr=12023#step:7:2875 is probably because of this **nor write** behavior. Previously there was one failing run and the bad caches would get overwritten. Now the bad caches stick there forever.

That's rational, but looks like the 

https://github.com/enso-org/enso/blob/cb72487cc962a4f3adc37845639363ecf847f4a4/engine/runtime-benchmarks/src/main/java/org/enso/compiler/benchmarks/module/ImportStandardLibrariesBenchmark.java#L129

may not be ready for that!

radeusgd replied:
Oh...

Can't we just run the engine with `--compile` to write the new caches, before running this bench?

JaroslavTulach commented on engine/runtime-compiler/src/main/scala/org/enso/compiler/context/InlineContext.scala:
```diff
@@ -18,22 +18,21 @@ import org.enso.compiler.pass.PassConfiguration
   */
 case class InlineContext(
-  private val module: ModuleContext,
+  moduleContext: ModuleContext,
```
It was `private`. Why open up?

radeusgd replied:
Hmmm good question. I think you are right this is unnecessary.

I think I needed it exposed when experimenting with something and then forgot to hide it again after the experiments turned out a dead-end. Sorry for the oversight and thanks for pointing it out.

JaroslavTulach commented on engine/runtime-integration-tests/src/test/java/org/enso/compiler/test/StaticAnalysisTest.java:
```diff
@@ -0,0 +1,117 @@
+   *
+   * <p>Ideally, the tests for the static analysis capabilities of the compiler should _not_ depend
+   * on the Graal runtime context, as they should be runnable in other contexts - i.e. in a Visual
+   * Studio Code language server.
```
That would indeed be ideal! The integration of the `runtime-compiler` into VSCode is blocked right now because we have no way to run the `runtime-compiler` without `TruffleCompilerContext` right now. If only these tests could be _moved to `runtime-compiler` project test_ directory! Then:
- they wouldn't depend on `TruffleCompilerContext`
- they would need an _independent compiler context_ implementation
- such an implementation could be used in VSCode as well

I remember @Akirathan was once dreaming about using `runtime-compiler` in VSCode and this was a huge _"known unknown"_.

radeusgd replied:
Yes I think ideally we should move them there - it seems like the right direction.

But implementing the independent compiler context is definitely a task for a separate PR.

JaroslavTulach commented on engine/runtime-integration-tests/src/test/java/org/enso/compiler/test/StaticAnalysisTest.java:
```diff
@@ -0,0 +1,117 @@
+    // If the module name is supposed to be put in a project, we register a synthetic project entry
+    // for it
+    if (qualifiedName.path().length() >= 2) {
+      LibraryName libraryName =
```
```suggestion
      var libraryName =
```

JaroslavTulach reviewed (approved):
I believe the abstraction of algorithms into _singletonizers_ and their usage is correct. I don't feel completely qualified to talk about the static analysis, but I left few comments inline. If the CI is green, then I am for integrating these changes.

radeusgd committed:
Merge branch 'develop' into wip/radeusgd/9812-infer-method-calls

radeusgd committed:
CR: naming, mark some classes final etc.

radeusgd committed:
fix after merge - change how to get instance of TruffleFileSystem

radeusgd committed:
CR: simplify how fallback scope is cached, make class static

radeusgd committed:
CR: rename materialized to resolved

radeusgd committed:
CR: use ScalaConversions

radeusgd committed:
CR: remove wrong FQN

radeusgd committed:
Merge branch 'develop' into wip/radeusgd/9812-infer-method-calls

radeusgd committed:
Simplify `BindingsMap` by instead relying on `StaticModuleScope` for atom type resolution (#11955)

radeusgd commented:
In [6b0ce25](https://github.com/enso-org/enso/pull/11399/commits/6b0ce2591d8d8ffa9c8ea0a02771fcc14b0fb839) I merged the followup PR so that we can merge them both to develop in one try.

I will proceed with merging once CI gets green.

radeusgd committed:
Merge branch 'develop' into wip/radeusgd/9812-infer-method-calls

radeusgd committed:
CR change var to interface type

radeusgd commented:
I've ran the benchmarks and tried comparing against develop:
[benchs.zip](https://github.com/user-attachments/files/18349627/benchs.zip)

It's really quite hard to tell in each benchmark if any deviations are just due to benchmark instability or real regression (how are you guys managing to live without seeing the standard deviation error bars on the measurements? 😮).

Here are a few 'more concerning' results I saw:
![image](https://github.com/user-attachments/assets/3e55e72f-0132-487b-915d-16a251d2cf26)
![image](https://github.com/user-attachments/assets/4328549e-a9b6-4ce9-9176-b5d9b9c960ff)
![image](https://github.com/user-attachments/assets/15a5771d-ec4c-490f-8bb6-af12839a8dc6)

Still they don't seem bad, so I think I will proceed with merging.

JaroslavTulach commented on engine/runtime-compiler/src/main/java/org/enso/compiler/pass/analyse/PassPersistance.java:
```diff
@@ -62,10 +64,12 @@
+@Persistable(clazz = TypeInferenceSignatures.class, id = 1281)
+@Persistable(clazz = FramePointerAnalysis$.class, id = 1282)
+@Persistable(clazz = TailCall.TailPosition.class, id = 1284)
+@Persistable(clazz = CachePreferences.class, id = 1285)
```
- @radeusgd, why have you added `CachePreferences` to this list?
- it is causing #11858

radeusgd replied:
I haven't **added** it. Please read the whole diff. It is been there before, I just moved it and changed the `id` (1284 -> 1285). Not sure why I did the id change, perhaps to keep them ordered. I guess the id change might have been unnecessary.

But `CachePreferences` were on that list before.

JaroslavTulach replied:
Right. I was confused. Your PR didn't introduce the persistance declaration.
